utopia-remix/app/store.tsx

```diff
@@ -1,44 +1,115 @@
+
+interface ProjectsStoreActions {
+  setSelectedProjectId: (projectId: string | null) => void
+  setSelectedCategory: (category: Category) => void
   setSearchQuery: (query: string) => void
-  sortCriteria: SortCriteria
   setSortCriteria: (sortCriteria: SortCriteria) => void
-  sortAscending: boolean
   setSortAscending: (sortAscending: boolean) => void
-  gridView: boolean
   setGridView: (gridView: boolean) => void
+  addOperation: (operation: Operation, key: string) => void
+  removeOperation: (key: string) => void
 }
 
+type ProjectsStore = ProjectsStoreState & ProjectsStoreActions
+
 export const useProjectsStore = create<ProjectsStore>()(
   devtools(
     persist(
       (set) => ({
-        selectedCategory: 'allProjects',
-        setSelectedCategory: (category: Category) => set(() => ({ selectedCategory: category })),
-        selectedProjectId: null,
-        setSelectedProjectId: (projectId: string | null) =>
-          set(() => ({ selectedProjectId: projectId })),
-        searchQuery: '',
-        setSearchQuery: (query) => set(() => ({ searchQuery: query })),
-        sortCriteria: 'dateModified',
-        setSortCriteria: (sortCriteria) => set(() => ({ sortCriteria: sortCriteria })),
-        sortAscending: false,
-        setSortAscending: (sortAscending) => set(() => ({ sortAscending: sortAscending })),
-        gridView: true,
-        setGridView: (gridView) => set(() => ({gridView: gridView})),
+        ...initialProjectsStoreStatePersisted,
+        ...initialProjectsStoreStateNonPersisted,
+
+        setSelectedCategory: (category: Category) => {
+          return set(() => ({ selectedCategory: category }))
+        },
+        setSelectedProjectId: (projectId: string | null) => {
+          return set(() => ({ selectedProjectId: projectId }))
+        },
+        setSearchQuery: (query) => {
+          return set(() => ({ searchQuery: query }))
+        },
+        setSortCriteria: (sortCriteria) => {
+          return set(() => ({ sortCriteria: sortCriteria }))
+        },
+        setSortAscending: (sortAscending) => {
+          return set(() => ({ sortAscending: sortAscending }))
+        },
+        setGridView: (gridView) => {
+          return set(() => ({ gridView: gridView }))
+        },
+        addOperation: (operation, key) => {
+          return set(({ operations }) => ({
+            operations: operations
+              .filter((other) => !operationsEqual(other, operation))
+              .concat(operationWithKey(operation, key)),
+          }))
+        },
+        removeOperation: (key) => {
+          return set(({ operations }) => ({
+            operations: operations.filter((other) => other.key !== key),
+          }))
+        },
       }),
       {
         name: 'store',
+        partialize: (fullStore) => {
+          const nonPersistedKeys = Object.keys(
+            initialProjectsStoreStateNonPersisted,
+          ) as (keyof ProjectsStoreStateNonPersisted)[]
```

Comment:
I suspect it's better to make this a type ascription, `as` will just tell TypeScript that it's definitely that type and if there's any divergence you'll have a bad time.

Reply:
I was having a hard time with the TS compiler around this one, do you have a direct example? (I thought this was fine since the keys of that object are definitely the keys of that type, but that's the best I got)

Reply:
I _believe_ this should work and will error if the types don't line up:
```typescript
const nonPersistedKeys: Array<keyof ProjectsStoreStateNonPersisted> = ...
```

Reply:
ah yeah, I tried it but…
```
Type 'string[]' is not assignable to type '(keyof ProjectsStoreStateNonPersisted)[]'.
  Type 'string' is not assignable to type 'keyof ProjectsStoreStateNonPersisted'.ts
```